### PR TITLE
[SPARK-19366][SQL] add getNumPartitions to Dataset

### PR DESCRIPTION
**Pull request by felixcheung**
## What changes were proposed in this pull request?

As suggested by @cloud-fan [here](https://github.com/apache/spark/pull/16668#discussion_r97254989), adding a simple wrapper in Scala can help avoid inefficiency with non-JVM cases

## How was this patch tested?

unit tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2421,6 +2421,13 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns the number of partitions of this Dataset.
+   * @group basic
+   * @since 2.2.0
+   */
+  def getNumPartitions: Int = rdd.getNumPartitions()
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: why is this not just numPartitions?

**Comment by rxin** [comment]
Actually - why do we need this? I worry it can be a confusing API due to optimizer behavior.

**Comment by felixcheung** [comment]
@rxin this was as suggested by @cloud-fan in more details in this thread [here](https://github.com/apache/spark/pull/16668#discussion_r97254989)

The original concerns were around the overhead with the extra conversion needed in Python and R (to PythonRDD, RRDD) and that it would be much lighter weight to have a method in Scala for this.

Now that we have a simple workaround in R (by just calling the Scala method without conversion), I'm not feeling strongly about this so I'm ok to close this.

I do agree with the optimizer behavior but this has been a very frequently requested method and its uses as `x.rdd.getNumPartitions` is all over PySpark code and documentation.

Perhaps it is worthwhile to explain this is a number to expect but can be optimized out.

**Comment by rxin** [comment]
Basically I want to push back against exposing this as a public API ...

**Comment by shivaram** [comment]
@rxin you think this will be confusing as the results might change over time ?

**Comment by rxin** [comment]
Yes.

**Comment by shivaram** [comment]
Isnt the dataset immutable ? i.e. the optimizer is called once when the RDD is materialized and the RDD doesn't change after that ?

**Comment by cloud-fan** [comment]
@shivaram yea, once the Dataset is materialized, the partition number won't change. I think rxin's concern is, when people trying to get the partition number, the result is unpredictable as too many factors can affect it.

**Pull request closed without merging**